gtk4-macros/src/util.rs

```diff
@@ -40,43 +43,76 @@ fn parse_attribute(meta: &NestedMeta) -> Result<(String, String)> {
         Some(ident) => ident,
     };
 
-    Ok((ident.to_string(), value))
+    Ok((ident.to_string(), ident.span(), value, meta.lit.span()))
 }
 
-fn parse_template_child_attribute(meta: &NestedMeta) -> Result<TemplateChildAttribute> {
-    let (ident, v) = parse_attribute(meta)?;
+fn parse_template_attribute(meta: &NestedMeta) -> Result<TemplateAttribute> {
+    let (ident, ident_span, value, value_span) = parse_attribute(meta)?;
 
     match ident.as_ref() {
-        "id" => Ok(TemplateChildAttribute::Id(v)),
-        s => bail!("Unknown item meta {}", s),
+        "id" => Ok(TemplateAttribute::Id(value, ident_span, value_span)),
+        "file" => Ok(TemplateAttribute::Filename(value, ident_span, value_span)),
+        s => abort!(ident_span, "Unknown template attribute `{}`", s),
     }
 }
 
-pub fn parse_template_child_attributes(
-    attr_name: &str,
-    attrs: &[Attribute],
-) -> Result<Vec<TemplateChildAttribute>> {
-    let meta_list = find_attribute_meta(attrs, attr_name)?;
-    let v = match meta_list {
-        Some(meta) => meta
-            .nested
-            .iter()
-            .map(|m| parse_template_child_attribute(&m))
+pub fn parse_attribute_iter(
+    meta_list: &mut dyn Iterator<Item = &NestedMeta>,
+) -> Result<Vec<TemplateAttribute>> {
+    Ok(Vec::from_iter(
+        meta_list
+            .map(|m| parse_template_attribute(&m))
             .fold_results(Vec::new(), |mut v, a| {
                 v.push(a);
                 v
             })?,
-        None => Vec::new(),
+    ))
+}
+
+pub fn parse_template_attributes(
+    attr_name: &str,
+    attrs: &[Attribute],
+) -> Result<Vec<TemplateAttribute>> {
+    let meta_list = find_attribute_meta(attrs, attr_name)?;
+    let v = match meta_list {
+        Some(meta) => parse_attribute_iter(&mut meta.nested.iter()),
+        None => Ok(Vec::new()),
     };
 
-    Ok(v)
+    v
 }
 
-pub fn crate_ident_new() -> Ident {
-    let crate_name = match crate_name("gtk4") {
+pub fn find_crate_name(name: &str) -> String {
+    match crate_name(name) {
         Ok(x) => x,
         Err(_) => "gtk4".to_owned(),
-    };
+    }
+}
+
+pub fn generate_type_from_class_name(class_name: &str) -> Result<Type, syn::Error> {
+    let mut ident_name = class_name.to_string();
 
-    Ident::new(&crate_name, Span::call_site())
+    if let Some(prefix) = class_name.get(0..3) {
+        if let Some(suffix) = class_name.get(3..) {
+            let mut lib = None;
+
+            if prefix.eq("Gtk") {
```

Comment:
Looks like this answers my question above - hardcoding the libraries here doesn't seem right, and won't work for custom widgets or different libraries.

Reply:
True. Not happy about it either. Is there a better way?

Reply:
How are Python/Vala/etc handling this? I guess the dynamic languages can just look into the typelibs at runtime to get the C type name mapping but Vala?

Reply:
Looking at vala it seems to build up a type registry at compile time. As in it registers all types in a dict to look up later. But not sure exactly, only glanced at the code.

Reply:
That would require us to read the .gir files at compile-time (slow, fragile) and also create a registry of all types inside the application somehow (I don't think that's possible unless we pre-process all code from build.rs).

Reply:
This is really difficult. The only way I could think of would be if all rust crates that provide widgets would alias their types somewhere so that no replacement needs to be done at all. gtk would for example alias `gtk::Box` as `GtkBox`. It could be done in the prelude or it could be in a separate module namespace like
```rust
use gtk::nativetypes;

// autogenerated
struct Widgets {
    label: nativetypes::GtkLabel;
}
```

If you had multiple libraries you would:
```rust
mod nativetypes {
    use gtk::nativetypes::*;
    use libhandy::nativetypes::*;
}

// autogenerated
struct Widgets {
    label: nativetypes::GtkLabel;
    action_row: nativetypes::HdyActionRow;
}